### PR TITLE
Make stealing more robust

### PR DESCRIPTION
**Pull request by hendrikmakait**
This PR improves a few details in the stealing code that could _potentially_ cause the scheduler to deadlock. I don't have any reproducer that could confirm this, but I've identified these weak points while investigating #8787. Fixing these should be strictly beneficial. 

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -569,7 +569,9 @@ def __hash__(self) -> int:
         return self._hash
 
     def __eq__(self, other: object) -> bool:
-        return isinstance(other, WorkerState) and other.server_id == self.server_id
+        return self is other or (
```
Comment (hendrikmakait) [inline distributed/scheduler.py:572]: This should speed up equality checks since we should mostly (always?) be using the same instance anyways.

Comment (fjetter) [reply distributed/scheduler.py:572]: should always be the same. I'd even be fine with removing `__eq__` entirely in which case `==` is the same as `is`

File: distributed/stealing.py [hunk 2]
```diff
@@ -178,13 +178,13 @@ def transition(
         *args: Any,
         **kwargs: Any,
     ) -> None:
-        if finish == "processing":
-            ts = self.scheduler.tasks[key]
-            self.put_key_in_stealable(ts)
-        elif start == "processing":
+        if start == "processing":
```
Comment (fjetter) [inline distributed/stealing.py:181]: Please put an in code comment here since this will otherwise die in a refactoring

**Review by fjetter — APPROVED** [review]

File: distributed/scheduler.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/scheduler.py:572]: > I'd even be fine with removing `__eq__` entirely in which case `==` is the same as `is`

I like the idea, but I'd leave it for a separate PR.

**Pull request merged**